### PR TITLE
Model id creation in service layer

### PR DESCRIPTION
**Pull request by tlm**
Reworks model creation id generation.
    
When creating models in DQlite via the service we use a
ModelCreationArgs struct that drives all of the meta data associated with
the model. Via previous changes we were putting the responsibility back
on to the caller of the create model process to generate model id.
    
This is an anti pattern that we want to remove entirely as it should the
services layer generating this and returning it to the caller. This
change the caller now receives the model id back as a return value.
    
There still exists two cases where the model id will be supplied
upfront. The first is via model migration where we are importing a model
that already exists and so we must use the id that has already been
established. The second is during a controller bootstrap when making the
controller model. This second case we can eventually get rid of and move
to the same pattern as the rest of the call sites but it will require
some big re-work to the Juju client at a later stage.
    
This commit also makes changes to secret back end selection for newly
created models. In previous commits a decision was made to pick a models
"default" secret back end in the sate layer. There was no reason to have
done this as all the context required is at the services layer. The code
has been moved one level up and removes the need to do more computation
inside of a transaction.
    
Renamed owner permission function for new models so that it now reflects
intent correctly. That is when creating a new model we wish to grant the
owner of the model admin permissions.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [x] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing
- ~[ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

This change is a re-work of what we already had. However while unit tests will be enough it would also be good to run a `juju bootstrap` and also add a few models and delete them.

## Documentation changes

N/A

### CODE REVIEW

**Review by manadart — COMMENTED** [review]
I am happy with adding a return for the UUID of the created model, but I would rather leave `UUID` in `ModelCreationArgs` and create one in the service layer if it is not supplied.

File: domain/model/bootstrap/bootstrap.go [hunk 1]
```diff
@@ -70,12 +89,25 @@ func CreateModel(
 				return fmt.Errorf("determining cloud type for model %q: %w", args.Name, err)
 			}
 
-			if err := state.Create(ctx, tx, modelType, args); err != nil {
-				return fmt.Errorf("create bootstrap model %q with uuid %q: %w", args.Name, args.UUID, err)
+			if args.SecretBackend == "" && modelType == coremodel.CAAS {
+				args.SecretBackend = kubernetessecrets.BackendName
+			} else if args.SecretBackend == "" && modelType == coremodel.IAAS {
+				args.SecretBackend = jujusecrets.BackendName
```
Comment (SimonRichardson) [inline domain/model/bootstrap/bootstrap.go:95]: OT: Nothing you've done, but those backend names should be centralised.

File: domain/model/service/service.go [hunk 2]
```diff
@@ -208,10 +273,51 @@ func (s *Service) CreateModel(
...
+// model with the same name and owner already exists.
+// - [errors.NotFound]: When the cloud, cloud region, or credential do not
+// exist.
+// - [github.com/juju/juju/domain/access/errors.NotFound]: When the owner of the
+// model can not be found.
+// - [modelerrors.AgentVersionNotSupported]: When the prescribed agent version
+// cannot be used with this controller or the agent version is set to zero.
+// - [secretbackenderrors.NotFound] When the secret backend for the model
+// cannot be found.
+func (s *Service) ImportModel(
```
Comment (SimonRichardson) [inline domain/model/service/service.go:300]: Thank-you for this, we've been fighting with create vs import for some time.

File: domain/model/bootstrap/bootstrap.go (hunk 1 above)
Comment (tlm) [reply domain/model/bootstrap/bootstrap.go:95]: @wallyworld this is one that you need to consider in your work please.

**Comment by tlm** [comment]
@manadart I understand what you have asked for but #17323 pulled this out because it was too confusing. So I have tried my best to get us back to where we were and address the confusion with the dual id that the comment wasn't solving.

**Review by hpidcock — APPROVED** [review]
Thank-you for pushing through with this.

File: domain/model/state/state.go [hunk 3]
```diff
@@ -105,36 +111,72 @@ func (s *State) Create(
...
+			input.Name, modelID, input.Owner, err,
+		)
 	}
 
-	if err := setModelSecretBackend(ctx, tx, input.UUID, modelType); err != nil {
-		return err
+	// Creates a record for the newly created model and register the target
+	// agent version.
+	if err := createModelAgent(ctx, tx, modelID, input.AgentVersion); err != nil {
+
```
[the remark this answers is not in the document]
Comment (tlm) [reply domain/model/state/state.go]: Fixed

File: domain/model/bootstrap/bootstrap.go [hunk 4]
```diff
@@ -70,12 +89,25 @@ func CreateModel(
 				return fmt.Errorf("determining cloud type for model %q: %w", args.Name, err)
 			}
 
-			if err := state.Create(ctx, tx, modelType, args); err != nil {
-				return fmt.Errorf("create bootstrap model %q with uuid %q: %w", args.Name, args.UUID, err)
+			if args.SecretBackend == "" && modelType == coremodel.CAAS {
```
Comment (wallyworld) [inline domain/model/bootstrap/bootstrap.go:92]: This should not be allowed to be empty. A secret backend is always needed to be specified. The default is "auto".

Comment (tlm) [reply domain/model/bootstrap/bootstrap.go:92]: This is exactly how it was implemented in the state layer. If this is wrong then I am going to push it back on yourself for the correct implementation.

**Pull request merged**